packages/checkout-storefront/src/lib/auth/SaleorAuthClient.ts

```diff
@@ -0,0 +1,190 @@
+import { SaleorAuthStorageHandler } from "./SaleorAuthStorageHandler";
+import { getRequestData, isExpiredToken } from "./utils";
+import {
+  CustomerDetachResponse,
+  CustomerDetachVariables,
+  Fetch,
+  PasswordResetResponse,
+  PasswordResetVariables,
+  TokenCreateResponse,
+  TokenCreateVariables,
+  TokenRefreshResponse,
+} from "./types";
+import invariant from "ts-invariant";
+import {
+  CHECKOUT_CUSTOMER_DETACH,
+  PASSWORD_RESET,
+  TOKEN_CREATE,
+  TOKEN_REFRESH,
+} from "@/checkout-storefront/lib/auth/mutations";
+
+export interface SaleorAuthClientProps {
+  onAuthRefresh?: (isAuthenticating: boolean) => void;
+  saleorApiUrl: string;
+  storage: Storage;
+}
+
+export class SaleorAuthClient {
+  private accessToken: string | null = null;
+  private tokenRefreshPromise: null | Promise<Response> = null;
+  private onAuthRefresh?: (isAuthenticating: boolean) => void;
+  private saleorApiUrl: string;
+  private storageHandler: SaleorAuthStorageHandler;
+  cleanup: SaleorAuthStorageHandler["cleanup"];
+
+  constructor({ saleorApiUrl, storage, onAuthRefresh }: SaleorAuthClientProps) {
+    this.storageHandler = new SaleorAuthStorageHandler(storage);
+    this.onAuthRefresh = onAuthRefresh;
+    this.saleorApiUrl = saleorApiUrl;
+    this.cleanup = this.storageHandler.cleanup;
+  }
+
+  private runAuthorizedRequest: Fetch = (input, init) => {
+    // technically we run this only when token is there
+    // but just to make typescript happy
+    if (!this.accessToken) {
+      return fetch(input, init);
+    }
+
+    const headers = init?.headers || {};
+
+    return fetch(input, {
+      ...init,
+      headers: { ...headers, Authorization: `Bearer ${this.accessToken}` },
+    });
+  };
+
+  private handleRequestWithTokenRefresh: Fetch = async (input, init) => {
+    const refreshToken = this.storageHandler.getRefreshToken();
+
+    invariant(refreshToken, "Missing refresh token in token refresh handler");
+
+    // the refresh already, finished, proceed as normal
+    if (this.accessToken) {
+      this.tokenRefreshPromise = null;
```

Comment:
Nah - had we removed it right away after the request is done and before the access token is set, there'd be a moment when another call could run that function, causing another refresh request because there'd be no access token and no refresh promise present at that moment. But I think we could move it to line 101, like so:
```
  this.accessToken = token;
  this.tokenRefreshPromise = null;
  return this.runAuthorizedRequest(input, init);
```